File: Problems/695/695_Max_Area_of_Island.py
from typing import List
from collections import deque

# bfs without using self
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROW = len(grid)
        COL = len(grid[0])
        directions = ((0, 1), (1, 0), (0, -1), (-1, 0))
        
        def bfs(r, c):
            if not(ROW>r>=0 and COL>c>=0)or not grid[r][c]:
                return 0
            grid[r][c] = 0
            total = 1
            qu = deque([(r, c)])
            while qu:
                length = len(qu)
                for _ in range(length):
                    r, c = qu.popleft()
                    for dr, dc in directions:
                        nr = r+dr
                        nc = c+dc
                        if ROW>nr>=0 and COL>nc>=0 and grid[nr][nc]==1:
                            qu.append((nr, nc))
                            grid[nr][nc] = 0
                            total += 1
            return total
        
        maxArea = 0
        for r in range(ROW):
            for c in range(COL):
                maxArea = max(maxArea, bfs(r, c))
        return maxArea

# bfs
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.curr_area = 0
        directions = ((1, 0), (0, 1), (-1, 0), (0, -1))
        ROW = len(grid)
        COL = len(grid[0])

        def bfs(r, c):
            if grid[r][c]==0:
                return
            grid[r][c] = 0
            self.curr_area += 1
            que = deque([[r,c]])
            while que:
                curr_r, curr_c = que.pop()
                for dir in directions:
                    dr, dc = dir
                    nxt_r = curr_r + dr
                    nxt_c = curr_c + dc
                    if ROW>nxt_r>=0 and COL>nxt_c>=0 and grid[nxt_r][nxt_c]:
                        que.append([nxt_r, nxt_c])
                        grid[nxt_r][nxt_c] = 0
                        self.curr_area += 1

        max_area = 0
        for r in range(ROW):
            for c in range(COL):
                self.curr_area = 0
                bfs(r, c)
                max_area = max(max_area, self.curr_area)
        return max_area

# dfs without self
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROW = len(grid)
        COL = len(grid[0])

        def dfs(r, c):
            if not (0 <= r < ROW and 0 <= c < COL):
                return 0
            if grid[r][c] != 1:
                return 0
            grid[r][c] = 0
            area = 1
            for dr, dc in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                nr = r + dr
                nc = c + dc
                area += dfs(nr, nc)
            return area

        res = 0
        for r in range(ROW):
            for c in range(COL):
                res = max(res, dfs(r, c))

        return res

# dfs
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.curr_area = 0
        ROW = len(grid)
        COL = len(grid[0])
        def dfs(r, c):
            if grid[r][c]:
                grid[r][c] = 0
                self.curr_area += 1
                if r-1>=0:
                    dfs(r-1, c)
                if ROW>r+1:
                    dfs(r+1, c)
                if c-1>=0:
                    dfs(r, c-1)
                if COL>c+1:
                    dfs(r, c+1)
        max_island = 0
        for r in range(ROW):
            for c in range(COL):
                self.curr_area = 0
                dfs(r, c)
                max_island = max(max_island, self.curr_area)
        return max_island

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = ((0, 1), (1, 0), (-1, 0), (0, -1))
        ROW = len(grid)
        COL = len(grid[0])
        def dfs(r, c):
            # Check if it is in boundary
            if not(0 <= r < ROW and 0 <= c < COL):
                return 0
            # Check if it is 1
            if grid[r][c] == 0:
                return 0
            grid[r][c] = 0
            cnt = 1
            for dr, dc in directions:
                cnt += dfs(r + dr, c + dc)
            return cnt
        res = 0
        for r in range(ROW):
            for c in range(COL):
                res = max(res, dfs(r, c))
        return res

File: Problems/695/695_Max_Area_of_Island.cpp
#include<vector>
#include<queue>

using namespace std;

// bfs without using class
class Solution {
public:
    vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_area = 0;
        for(int r=0; r<grid.size(); r++){
            for(int c=0; c<grid[0].size(); c++){
                max_area = max(max_area, bfs(grid, r, c));
            }
        }
        return max_area;
    }
    
    int bfs(vector<vector<int>> &grid, int r, int c){
        if(!grid[r][c]) return 0;
        queue<pair<int, int>> qu;
        qu.push({r, c});
        grid[r][c] = 0;
        int total=1;
        while(!qu.empty()){
            int curr_r = qu.front().first;
            int curr_c = qu.front().second;
            qu.pop();
            for(auto dir: directions){
                int nxt_r = curr_r + dir.first;
                int nxt_c = curr_c + dir.second;
                if(grid.size()>nxt_r && nxt_r>=0 && grid[0].size()>nxt_c && nxt_c>=0 && grid[nxt_r][nxt_c]){
                    qu.push({nxt_r, nxt_c});
                    grid[nxt_r][nxt_c] = 0; 
                    total++;
                }
            }
        }
        return total;
    }
};

// bfs
class Solution {
public:
    int curr_area = 0;
    vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_area = 0;
        for(int r=0; r<grid.size(); r++){
            for(int c=0; c<grid[0].size(); c++){
                curr_area = 0;
                bfs(grid, r, c);
                max_area = max(max_area, curr_area);
            }
        }
        return max_area;
    }
    void bfs(vector<vector<int>> &grid, int r, int c){
        if(!grid[r][c]) return;
        curr_area++;
        queue<pair<int, int>> qu;
        qu.push({r, c});
        grid[r][c] = 0;
        while(!qu.empty()){
            int curr_r = qu.front().first;
            int curr_c = qu.front().second;
            qu.pop();
            for(auto dir: directions){
                int nxt_r = curr_r + dir.first;
                int nxt_c = curr_c + dir.second;
                if(grid.size()>nxt_r && nxt_r>=0 && grid[0].size()>nxt_c && nxt_c>=0 && grid[nxt_r][nxt_c]){
                    qu.push({nxt_r, nxt_c});
                    grid[nxt_r][nxt_c] = 0; 
                    curr_area++;
                }
            }
        }
    }
};

// dfs without using class parameters
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int maxArea = 0;
        for(int r=0; r<grid.size(); r++){
            for(int c=0; c<grid[0].size(); c++){
                maxArea = max(maxArea, dfs(r, c, grid));
            }
        }
        return maxArea;
    }
    int dfs(int r, int c, vector<vector<int>>& grid){
        if(r>=grid.size() || 0>r || c>=grid[0].size() || 0>c || !grid[r][c]) return 0;
        int total = 1;
        grid[r][c] = 0;
        vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for(auto dir: directions){
            total += dfs(r+dir.first, c+dir.second, grid);
        }
        return total;
    }
};

// dfs
// class Solution {
// public:
//     int curr_area = 0;
//     int maxAreaOfIsland(vector<vector<int>>& grid) {
//         int max_area = 0;
//         for(int r=0; r<grid.size(); r++){
//             for(int c=0; c<grid[0].size(); c++){
//                 curr_area = 0;
//                 dfs(grid, r, c);
//                 max_area = max(max_area, curr_area);
//             }
//         }
//         return max_area;
//     }

//     void dfs(vector<vector<int>>& grid, int r, int c){
//         if(grid[r][c]){
//             grid[r][c] = 0;
//             curr_area++;
//             if(r-1>=0)          dfs(grid, r-1, c);
//             if(grid.size()>r+1) dfs(grid, r+1, c);
//             if(c-1>=0)          dfs(grid, r, c-1);
//             if(grid[0].size()>c+1) dfs(grid, r, c+1);
//         }
//     }
// };